[ZIP 233] NSM: Burning (#913)
zcash/zips: zips/zip-0233.md

and "REQUIRED" in this document are to be interpreted as described in RFC 2119.
[1]

The term "network upgrade" in this document is to be interpreted as described in
ZIP 200. [2]

"Block Subsidy” - the algorithmic issuance of ZEC on block creation. Part of the
consensus rules. Split between the miner and the Dev Fund. Also known as Block
Reward.

"Issuance" - The method by which ZEC becomes available for circulation on the
network.

"We" - the ZIP Owners and Authors, listed in the above front matter.

"`MAX_MONEY`" is the total ZEC supply cap, defined as 21,000,000 ZEC. This is
slightly larger than the supply cap for the current issuance mechanism, but is
the value used in existing critical consensus checks.

# Abstract

We propose the introduction of a mechanism to voluntarily burn funds, removing
those funds entirely from circulation on the network. This mechanism, in
combination with ZIPs 234 and 235, comprises a long-term strategy for the
sustainability of the network. We will refer to the combined effects of these

> three ZIPs as the “Network Sustainability Mechanism”.

Reviewer: @nuttycom, @str4d, and I think that this may not be a well-justified change in terminology, because:
* in other cryptocurrencies like Bitcoin, burnt funds can *never* be reissued;
* the term "burning" may conjure up associations that are not accurate in this context.

"Unissuance" seems fine to us.

Author: > in other cryptocurrencies like Bitcoin, burnt funds can never be reissued

But that _is_ also the intention here. It seems there could be legal complications with the idea of funds being somehow held in trust or otherwise kept and re-issued. So the intention is to permanently destroy the funds, and allow the economy to react. If this ZIP were taken in isolation, the only effect would be deflation, similarly to Ethereum. But in the context of ZIP-234, which redefines the issuance curve in terms of the chain value, the effect of reducing the chain value takes on additional significance.

# Specification

The modifications required are:

1. The addition of a transaction field representing an amount to burn for a
   given transaction.
2. The inclusion of the burn amount in the calculated total output value for a
   given transaction.
3. Consensus rules to ensure the burned amount is no longer available for
   circulation on the network.

## Burning

We add a new field to the common transaction fields [3]:

    burnAmount : uint64 [zatoshi]

This field specifies how much of the input value for the transaction should be
burned.

**Note:** Older transaction versions can continue to be supported after a
network upgrade, but burning is not possible for these transactions. For
example, NU5 supports both v4 and v5 transaction formats, for both coinbase and
non-coinbase transactions.

## Consensus Rule Changes

The burned value must now be considered when calculating the total output value
of a transaction. It should be treated similarly to a transparent output,
except that there is no way for this value to be used as an input in a future
transaction.

## Transaction Format Requirements

* ZIP 230 [3] must be updated to include the new `burnAmount` field.

> * ZIP 244 [4] must be updated as well to include the new `burnAmount` field.

Reviewer: This depends on whether the v6 sighash changes are done as a modification to ZIP 244 or as a new ZIP.
```suggestion
* The `burnAmount` field must be taken account of in sighash computation
  (either in ZIP 244 [4] or in a new ZIP that specifies the sighash algorithm
  for the new transaction version).
```

Reviewer 2: Please follow the example in https://zips.z.cash/draft-ecc-explicit-fees to specify the changes to the transaction format.

Author: I've tried to apply this in this commit: https://github.com/zcash/zips/pull/913/commits/86c35a9c5a3924d7eb2f9b2b8148aea86d94cdaf.

# Rationale

All technical decisions in this ZIP are balanced between the necessary
robustness of the NSM mechanics, and simplicity of implementation.

## New transaction field for burn amount

> An explicit value distinguishes the burned ZEC from the implicit transaction
> fee. Explicitness also ensures any arithmetic flaws in any implementations are
> more likely to be observed and caught immediately.

Reviewer: It is unlikely that the transaction fee will remain implicit in the new transaction version. The current implicitness of the fee is also not relevant to the point being made; it's sufficient to say that the burned value needs to be distinguished from the fee, whether the latter is implicit or explicit.
```suggestion
An explicit value distinguishes the burned ZEC from the transaction fee.
Explicitness also ensures any arithmetic flaws in implementations are more
likely to be observed and caught immediately.
```

Author: Applied in [this commit](https://github.com/zcash/zips/pull/913/commits/fad4eae55eecb19a19dcac999c2b2a214887b596) :heavy_check_mark:

**Note:** Older transaction versions can continue to be supported after a
network upgrade, but burning is not possible for these transactions. For
example, NU5 supports both v4 and v5 transaction formats, for both coinbase and
non-coinbase transactions.

## Consensus Rules

The burned value must now be considered when calculating the total output value
of a transaction. It should be treated similarly to a transparent output,
except that there is no way for this value to be used as an input in a future
transaction.

## Digests

The transaction digest algorithm defined in ZIP 244 [4] is to be modified for v6
transactions as follows:

Section T.1: header_digest [5] is specified in draft-txv6-sighash  [6] to read:

> A BLAKE2b-256 hash of the following values:
>
> ```
> T.1a: version             (4-byte little-endian version identifier including ``fOverwintered`` flag)
> T.1b: version_group_id    (4-byte little-endian version group identifier)
> T.1c: consensus_branch_id (4-byte little-endian consensus branch id)
> T.1d: lock_time           (4-byte little-endian ``nLockTime`` value)
> T.1e: expiry_height       (4-byte little-endian block height)

> > T.1f: burn_amount         (8-byte little-endian burn amount value)

Reviewer: This has an interaction with adding the explicit fee field, but the intent is clear.

sustainability of the network. We will refer to the combined effects of these
three ZIPs as the “Network Sustainability Mechanism”.

# Motivation

This mechanism seeks to address concerns about the sustainability of the network
design shared by Bitcoin-like systems:

1. **Long Term Consensus Sustainability:** By enabling the burning of funds, the
   network gains the ability to create "headroom" between the chain value and
   `MAX_MONEY`. This lays necessary groundwork for extending the miner reward
   system, which currently has a clear final end date.
2. **Benefits to ZEC Holders:** Burning funds reduces the supply of ZEC,
   benefiting network users in proportion to their holdings without requiring
   them to opt into any scheme, introducing extra risk, active oversight, or
   accounting complexity.

# Specification

The modifications required are:

1. The addition of a transaction field representing an amount to burn for a
   given transaction.
2. The inclusion of the burn amount in the calculated total output value for a
   given transaction.
3. Consensus rules to ensure the burned amount is no longer available for
   circulation on the network.

## Transaction Format

> The following field is added to the v7 transaction format [3]:

Reviewer 2: ```suggestion
The following field is added to the v6 transaction format [3]:
```

Author: Fixed ✅

> The following field is added to the v7 transaction format [3]:

Reviewer: ```suggestion
The following field is added to the v6 transaction format [3]:
```

Author: Fixed ✅

+-------+----------------+------------+------------------------------------------------------+
| Bytes | Name           | Data Type  | Description                                          |
+=======+================+============+======================================================+
|   8   | ``burnAmount`` | ``uint64`` | The amount of input value to be burned, in zatoshis. |
+-------+----------------+------------+------------------------------------------------------+

**Note:** Older transaction versions can continue to be supported after a
network upgrade, but burning is not possible for these transactions. For
example, NU5 supports both v4 and v5 transaction formats, for both coinbase and
non-coinbase transactions.

## Consensus Rules

The burned value must now be considered when calculating the total output value
of a transaction. It should be treated similarly to a transparent output,
except that there is no way for this value to be used as an input in a future

> transaction.

Reviewer 2: Out interpretation of this (from which we will specify the actual consensus rules that replace this text during ZIP editing) is:

The burned value subtracts from the remaining value in the "transparent transaction value pool" as described in section 3.4: Transactions and Treestates and does not result in an output being produced in any chain value pool.

Author: This looks correct to me, yes. To clarify the intention: it should be possible to burn funds that originate from shielded inputs too. I _think_ this is included in the definition of "transparent transaction value pool"?

Reviewer 2: That's correct, the "transparent transaction value pool" is _all_ value that is visible to chain observers; this includes the various valueBalance* fields and of course the explicitly transparent value.

## Digests

> The transaction digest algorithm defined in ZIP 244 [4] is to be modified for v7

Reviewer 2: ```suggestion
The transaction digest algorithm defined in ZIP 244 [4] is to be modified for v6
```

> The transaction digest algorithm defined in ZIP 244 [4] is to be modified for v7

Reviewer: ```suggestion
The transaction digest algorithm defined in ZIP 244 [4] is to be modified for v6
```

transactions as follows:

A section is inserted, specified to read:

> Section T.5: `burn_amount_digest`
>
> A BLAKE2b-256 hash of the following values:
>
> ```
> T.5a: burnAmount (8-byte little-endian value)
> ```
>
> The personalization field of this hash is set to:
>
> ```
> ZTxBurnAmnt_Hash

> > ```

Reviewer 2: This will instead be included in the header digest; it does not require a separate section. If there is a specific rationale for including this as an additional section instead, please state that rationale.

Author: This sounds OK to me. I'll adjust the ZIP and the implementation PRs :+1:

Author: Fixed with this diff: https://github.com/zcash/zips/compare/d590947513d9c8cb2c5e772ecacf56c8b3cd1246..7e0431dc58e6236d9b7b051dc020435075780899

`txid_digest` is to be modified to append T.5 (`burn_amount_digest`) to the list
of values.

> The signature digest algorithm defined in ZIP 244 [4] is to be modified for v7

Reviewer: ```suggestion
The signature digest algorithm defined in ZIP 244 [4] is to be modified for v6
```

Author: Fixed ✅

transactions as follows:

A section is inserted, specified to read:

> Section S.5: `burn_amount_digest`
>
> A BLAKE2b-256 hash of the following values:
>
> ```
> S.5a: burnAmount (8-byte little-endian value)
> ```
>
> The personalization field of this hash is set to:
>
> ```
> ZTxBurnAmnt_Hash

> > ```

Reviewer 2: Again, this will also be part of the header digest (unless there is a specific rationale for a separate section.)

Author: Will do :+1:

> > ```

Reviewer: This only needs to say that the `burnAmount` field is committed to in the hash of the transaction header; nothing more.

Author: Will do :+1: I'll resolve this and track the change in @nuttycom's comment thread to avoid duplication.
